### PR TITLE
[BREAKING][dist, model, trainer] refactor: make parallel state model-owned

### PR DESCRIPTION
**Pull request by Luosuu**
[BREAKING][dist, model, trainer] refactor: make parallel state model-owned

### What does this PR do?

Makes `ParallelState` model-owned so one RL trainer can coordinate policy, reference, reward, and drafter models with different parallel topologies.

This refines the design discussed in #893: model binding is the source of truth, while a `ContextVar` scopes existing patched modeling code that calls `get_parallel_state()` to the active model execution.

### Checklist Before Starting

- Related PR: #893
- PR title follows `[{modules}] {type}: {description}` format.

### Test

- `source .venv/bin/activate && make quality`
- `python3 -m compileall -q veomni tasks tests`
- Added model-ownership, nested-context, worker-pickle, process-group-lifetime, fullgraph compile, strict group-resolution, test-group-precedence, and autograd backward regression tests.
- Full pytest cannot run in the local macOS environment because the project GPU/NPU PyTorch stack is unavailable there. GPU, NPU, distributed, and torch.compile validation is delegated to the GitHub CI pipeline.

### API and Usage Example

```python
policy_state = build_parallel_state(dp_size=8, ulysses_size=1)
drafter_state = build_parallel_state(dp_size=4, ulysses_size=2)

policy = build_foundation_model(..., parallel_state=policy_state)
drafter = build_foundation_model(..., parallel_state=drafter_state)

policy = build_parallelize_model(policy, ...)
drafter = build_parallelize_model(drafter, ...)

draft = drafter(**draft_inputs)
with use_model_parallel_state(policy):
    outputs = policy(**policy_inputs)
    outputs.loss.backward()
```

This is a breaking API change: `init_parallel_state()` and the process-global default state are removed. Callers must retain `build_parallel_state()`'s result and pass it explicitly.

### Design & Code Changes

- Add side-effect-free `build_parallel_state()` with process-group-lifetime-aware topology caching.
- Bind a non-serialized state reference in `build_parallelize_model(..., parallel_state=...)`.
- Keep bootstrap state local to model construction; trainers do not retain a `parallel_state` attribute.
- Add nested and exception-safe model execution contexts backed by `ContextVar`.
- Mark context lookups as per-graph constants so `torch.compile(fullgraph=True)` captures each model's topology without tracing `ContextVar.get()`.
- Resolve optimizer, gradient clipping, checkpoint, collator, and dataloader behavior from explicit/model-owned state.
- Make DPO policy/reference and DiT condition/foundation execution select state by model.
- Resolve SP communication groups only from the active model context, with a narrow Ulysses group override for standalone communication tests.
- Keep worker collators pickle-safe by storing only scalar SP metadata.
- Capture SP world size during `ReduceLoss.forward()` so backward never reads another model's ambient state.
- Update developer constraints, architecture notes, and user documentation.

### Checklist Before Submitting

- [x] Read the Contribute Guide
- [x] Applied pre-commit quality checks
- [x] Added/updated documentation
- [x] Added tests for the new ownership and lifecycle behavior

### CODE REVIEW

**Comment by Coach257** [comment]
#893
